src/app/qgsapplayertreeviewmenuprovider.cpp

```diff
@@ -247,64 +247,109 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
-                  if ( sqlLayer->isValid() )
-                  {
-                    layer->setDataSource( sqlLayer->source(), sqlLayer->name(), sqlLayer->dataProvider()->name(), QgsDataProvider::ProviderOptions() );
-                    queryResultWidget->notify( QObject::tr( "Layer Update Success" ), QObject::tr( "The SQL layer was updated successfully" ), Qgis::MessageLevel::Success );
-                  }
-                  else
+                  try
                   {
-                    QString error { sqlLayer->dataProvider()->error().message( QgsErrorMessage::Format::Text ) };
-                    if ( error.isEmpty() )
+                    std::unique_ptr<QgsMapLayer> sqlLayer { conn3->createSqlVectorLayer( options ) };
+                    if ( sqlLayer->isValid() )
                     {
-                      error = QObject::tr( "layer is not valid, check the log messages for more information" );
+                      layer->setDataSource( sqlLayer->source(), sqlLayer->name(), sqlLayer->dataProvider()->name(), QgsDataProvider::ProviderOptions() );
+                      queryResultWidget->notify( QObject::tr( "Layer Update Success" ), QObject::tr( "The SQL layer was updated successfully" ), Qgis::MessageLevel::Success );
                     }
-                    queryResultWidget->notify( QObject::tr( "Layer Update Error" ), QObject::tr( "Error updating the SQL layer: %1" ).arg( error ), Qgis::MessageLevel::Critical );
+                    else
+                    {
+                      QString error { sqlLayer->dataProvider()->error().message( QgsErrorMessage::Format::Text ) };
+                      if ( error.isEmpty() )
+                      {
+                        error = QObject::tr( "layer is not valid, check the log messages for more information" );
+                      }
+                      queryResultWidget->notify( QObject::tr( "Layer Update Error" ), QObject::tr( "Error updating the SQL layer: %1" ).arg( error ), Qgis::MessageLevel::Critical );
+                    }
+                  }
+                  catch ( QgsProviderConnectionException &ex )
+                  {
+                    queryResultWidget->notify( QObject::tr( "Layer Update Error" ), QObject::tr( "Error updating the SQL layer: %1" ).arg( ex.what() ), Qgis::MessageLevel::Critical );
                   }
                 }
-                catch ( QgsProviderConnectionException &ex )
+
+              } );
+
+              dialog.exec();
+
+            }
+          } );
+
+          // SQL dialog
+          if ( conn->capabilities().testFlag( QgsAbstractDatabaseProviderConnection::Capability::ExecuteSql ) )
+            menu->addAction( QgsApplication::getThemeIcon( QStringLiteral( "/dbmanager.svg" ) ), tr( "Execute SQL…" ), menu, [ layer, this ]
+          {
+            std::unique_ptr< QgsAbstractDatabaseProviderConnection> conn2 { QgsMapLayerUtils::databaseConnection( layer ) };
```

Comment:
For safety, I think `layer` should by made a QPointer so that we don't risk trying to access a removed layer here. (The event loop have had a chance to run since the previous nullptr checks on layer/vlayer)